[quant][graph] Add valueQuantizable function

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36622 [quant][graph] Graph mode quantization support for sigmoid
* #36552 [quant][graph] Add quantized batch_norm2d_relu to graph mode
* #36549 [quant][graph] Add quantized batch_norm2d support to graph mode
* **#36548 [quant][graph] Add valueQuantizable function**

Summary:
Refactor to be able to observe based on inputs to ops

Test Plan:
test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21048963](https://our.internmc.facebook.com/intern/diff/D21048963)

supriyar committed:
[quant][graph] Add valueQuantizable function
Summary:
Refactor to be able to observe based on inputs to ops

Test Plan:
test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -986,10 +986,20 @@ void InsertObserversHelper::preprocess(
-bool isDynamicLSTMWeight(Value* v, Use use, bool is_dynamic) {
-  return is_dynamic && use.user->kind() == Symbol::aten("lstm") &&
-      (use.offset == 2);
+bool valueQuantizable(Use use, bool is_dynamic) {
```
since this is checking the use, could you rename it to `useQuantizable`?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -986,10 +986,20 @@ void InsertObserversHelper::preprocess(
+bool valueQuantizable(Use use, bool is_dynamic) {
+  Node* n = use.user;
+  // static quantization
+  if (n->kind() == prim::CallFunction &&
+      getFuncName(n->inputs()[0]) == "batch_norm") {
+    return use.offset == 1;
+  }
```
could you refactor this further to take a list of tuple of (name, offset) and check if the use is matching the name and offset?

jerryzh168 replied:
some similar functions we already implemented: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization.cpp#L707, https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization.cpp#L713, please feel free to refactor these as well as you see fit

supriyar replied:
I've added the list of tuple in the function itself. I didn't want to spend too much time refactoring the other code, so I used it in the best way I saw fit, given that the conditions we were returning on in both cases was different.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -986,10 +986,20 @@ void InsertObserversHelper::preprocess(
+  }
+  // Dynamic quantization checks
+  if (is_dynamic) {
+    if (use.user->kind() == Symbol::aten("lstm")) {
+      return use.offset == 2;
+    }
```
similar for this one

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -986,10 +986,20 @@ void InsertObserversHelper::preprocess(
+      return use.offset == 2;
+    }
+  }
+  return nodeQuantizable(n);
```
I feel hiding this here is not very explicit, I think moving this outside and use `nodeQuantizable && useQuantizable` is probably better

supriyar committed:
Update on "[quant][graph] Add valueQuantizable function"


Summary:
Refactor to be able to observe based on inputs to ops

Test Plan:
test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -986,10 +987,25 @@ void InsertObserversHelper::preprocess(
+
+  const AtenFuncArgs& aten_func_args = AtenFuncArgs({{"lstm", 2}});
+  const CallFuncArgs& call_func_args = CallFuncArgs({{"batch_norm", 1}});
+  for (const auto& func_arg : aten_func_args) {
+    if (n->kind() == Symbol::aten(func_arg.func_name)) {
+      return v == n->inputs().at(func_arg.arg_index);
+    }
+  }
+
+  for (const auto& func_arg : call_func_args) {
+    if (n->kind() == prim::CallFunction &&
+        getFuncName(n->inputs()[0]) == func_arg.func_name) {
+      return v == n->inputs().at(func_arg.arg_index);
+    }
+  }
```
I think you can use isAtenFuncNthArg and isCallFunctionNthArg here, also you can refactor these functions to not take value as input, and just compare `use.offset` with the `arg_index`

jerryzh168 replied:
the logic seems a bit different here actually, and this is due to the convoluted check logic that we fall back to nodeQuantizable when the check fails. Can we add a default index for aten and call function in nodes in nodeQuantizable?

supriyar committed:
Update on "[quant][graph] Add valueQuantizable function"


Summary:
Refactor to be able to observe based on inputs to ops

Test Plan:
test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -675,24 +676,28 @@ graph(%self, %a, %b, %inplace):
-bool isAtenFuncNthArg(
-    Value* v,
+// `v` is the nth argument (if provided) of the function.
+bool matchAtenFuncToUse(
```
Can we remove `Node*` and `Value` and just have `Use` here?

supriyar replied:
Isn't the value necessary to compare this `v == use->inputs().at(n)`?

jerryzh168 replied:
no, you can do use.offset == n

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -675,24 +676,28 @@ graph(%self, %a, %b, %inplace):
-bool isCallFunctionNthArg(
-    Value* v,
+// `v` is the nth argument (if provided) of the function
+bool matchCallFuncToUse(
```
same for this one

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -986,10 +992,30 @@ void InsertObserversHelper::preprocess(
+  // For each operator in this list observers are inserted for the input based
+  // on the index specified.
+  const AtenFuncArgs& aten_func_args = AtenFuncArgs({{"lstm", 2}});
+  const CallFuncArgs& call_func_args = CallFuncArgs({{"batch_norm", 1}});
```
this change belongs to next PR

supriyar committed:
Update on "[quant][graph] Add valueQuantizable function"


Summary:
Refactor to be able to observe based on inputs to ops

Test Plan:
test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add valueQuantizable function"


Summary:
Refactor to be able to observe based on inputs to ops

Test Plan:
test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -675,24 +676,26 @@ graph(%self, %a, %b, %inplace):
-      getFuncName(use->inputs()[0]) == func_name && v == use->inputs().at(n);
+    c10::optional<int> n) {
+  Node* node = use.user;
+  return node->kind() == prim::CallFunction &&
```
false --> true?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -986,10 +988,29 @@ void InsertObserversHelper::preprocess(
+  // Special checks for ops that do not require observers for all input tensors.
+  // For each operator in this list observers are inserted for the input based
+  // on the index specified.
+  const AtenFuncArgs& aten_func_args = AtenFuncArgs({{"lstm", 2}});
```
we also need to check `is_dynamic` for "lstm" node..

supriyar replied:
I think that won't be necessary because for dynamic quant we insert observers at the input. Which should be same as static quant. We have a check for `is_dynamic` to insert observer at the output of node in the parent function.

jerryzh168 replied:
OK, then do you need `is_dynamic` as argument here?

jerryzh168 replied:
we don't support static quant right now though

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -986,10 +988,29 @@ void InsertObserversHelper::preprocess(
-  return is_dynamic && use.user->kind() == Symbol::aten("lstm") &&
-      (use.offset == 2);
+bool useQuantizable(const Use& use, bool is_dynamic) {
+  bool result = nodeQuantizable(use.user);
+  if (!result) {
+    return false;
+  }
```
move this to the end of the function?

jerryzh168 replied:
- just return nodeQUantizable(use.user) in the end

supriyar committed:
Update on "[quant][graph] Add valueQuantizable function"


Summary:
Refactor to be able to observe based on inputs to ops

Test Plan:
test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -986,10 +988,31 @@ void InsertObserversHelper::preprocess(
+    }
+  }
+  // Dynamic quantized ops that require special handling for inputs.
+  if (is_dynamic) {
+    if (matchAtenFuncToUse(use, "lstm", c10::nullopt)) {
```
nit: merge the conditions

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -986,10 +988,31 @@ void InsertObserversHelper::preprocess(
+  // Special checks for ops that do not require observers for all input tensors.
+  // For each operator in this list observers are inserted for the input based
+  // on the index specified.
+  const AtenFuncArgs& aten_func_args = AtenFuncArgs({});
+  const CallFuncArgs& call_func_args = CallFuncArgs({});
```
could you expose these two lists in the beginning of the file so that it's easier to find and change?

jerryzh168 replied:
Also, a follow up here is, can we remove the isBias check for conv and linear and put the special case in the list?

supriyar replied:
> Also, a follow up here is, can we remove the isBias check for conv and linear and put the special case in the list?

I think we can do this separately, I also see `isWeight` check that uses these lists.

jerryzh168 reviewed (approved):
Looks good, thanks! left a few more comments, please address them before landing

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -675,24 +676,26 @@ graph(%self, %a, %b, %inplace):
+  Node* node = use.user;
+  return node->kind() == prim::CallFunction &&
+      getFuncName(node->inputs()[0]) == func_name &&
+      (n.has_value() ? (n.value() == use.offset) : true);
```
actually this can be changed to `!n.has_value() || n.value() == use.offset`

supriyar committed:
Update on "[quant][graph] Add valueQuantizable function"


Summary:
Refactor to be able to observe based on inputs to ops

Test Plan:
test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -110,6 +111,20 @@ std::vector<std::string> _single_input_general_aten_funcs = {
+// Special checks for ops that do not require observers for all input tensors.
+// For each operator in this list observers are inserted for the input based
+// on the index specified.
+AtenFuncArgs _aten_func_args = {};
+CallFuncArgs _call_func_args = {};
```
can you use better names for these?
